File: src/org/simbrain/network/neuron_update_rules/DecayRule.java
/*
 * Part of Simbrain--a java-based neural network kit
 * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package org.simbrain.network.neuron_update_rules;

import org.simbrain.network.core.Network.TimeType;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.NeuronUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.BoundedUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.ClippableUpdateRule;
import org.simbrain.network.neuron_update_rules.interfaces.NoisyUpdateRule;
import org.simbrain.util.UserParameter;
import org.simbrain.util.math.ProbDistributions.UniformDistribution;
import org.simbrain.util.math.ProbabilityDistribution;

/**
 * <b>DecayNeuron</b> implements various forms of standard decay.
 */
public class DecayRule extends NeuronUpdateRule implements BoundedUpdateRule, ClippableUpdateRule, NoisyUpdateRule {

    /**
     * The Default upper bound.
     */
    private static final double DEFAULT_CEILING = 1.0;

    /**
     * The Default lower bound.
     */
    private static final double DEFAULT_FLOOR = -1.0;

    /**
     * Relative.
     */
    public static final int RELATIVE = 0;

    /**
     * Absolute.
     */
    public static final int ABSOLUTE = 1;

    //TODO: add drop down back
    /**
     * Relative absolute.
     */
    private int relAbs = RELATIVE;

  //TODO: disable when RELATIVE
    /**
     * Decay amount.
     */
    @UserParameter(
            label = "Decay amount",
            description = "The amount by which the activation is changed each iteration if absolute decay is chosen.",
            defaultValue = "0.1", order = 3)
    private double decayAmount = .1;

    //TODO: disable when ABSOLUTE
    /**
     * Decay fraction.
     */
    @UserParameter(
            label = "Decay fraction",
            description = "The proportion of the distance between the current value and the base-line value, "
                    + "by which the activation is changed each iteration if relative decay is chosen.",
            defaultValue = "0.1", order = 4)
    private double decayFraction = .1;

    /**
     * Base line.
     */
    @UserParameter(
            label = "Base Line",
            description = "An option to add noise.",
            defaultValue = "0", order = 2)
    private double baseLine = 0;

    /**
     * Clipping.
     */
    private boolean clipping = true;

    /**
     * Noise generator.
     */
    private ProbabilityDistribution noiseGenerator = UniformDistribution.create();

    /**
     * Add noise to the neuron.
     */
    private boolean addNoise = false;

    /**
     * The upper bound of the activity if clipping is used.
     */
    private double ceiling = DEFAULT_CEILING;

    /**
     * The lower bound of the activity if clipping is used.
     */
    private double floor = DEFAULT_FLOOR;

    @Override
    public void update(Neuron neuron) {
        double val = neuron.getActivation() + neuron.getInput();
        double decayVal = 0;

        if (relAbs == RELATIVE) {
            decayVal = decayFraction * Math.abs(val - baseLine);
        } else if (relAbs == ABSOLUTE) {
            decayVal = decayAmount;
        }

        // Here's where the action happens
        if (val < baseLine) {
            val += decayVal;

            // in case of an overshoot
            if (val > baseLine) {
                val = baseLine;
            }
        } else if (val > baseLine) {
            val -= decayVal;

            // in case of an overshoot
            if (val < baseLine) {
                val = baseLine;
            }
        }

        if (addNoise) {
            val += noiseGenerator.getRandom();
        }

        if (clipping) {
            val = clip(val);
        }

        neuron.setBuffer(val);
    }


    @Override
    public TimeType getTimeType() {
        return TimeType.DISCRETE;
    }

    @Override
    public DecayRule deepCopy() {
        DecayRule dn = new DecayRule();
        dn.setRelAbs(getRelAbs());
        dn.setDecayAmount(getDecayAmount());
        dn.setDecayFraction(getDecayFraction());
        dn.setClipped(isClipped());
        dn.setUpperBound(getUpperBound());
        dn.setLowerBound(getLowerBound());
        dn.setAddNoise(getAddNoise());
        dn.noiseGenerator = noiseGenerator.deepCopy();
        return dn;
    }

    @Override
    public double clip(double val) {
        if (val > getUpperBound()) {
            return getUpperBound();
        } else if (val < getLowerBound()) {
            return getLowerBound();
        } else {
            return val;
        }
    }

    @Override
    public void contextualIncrement(Neuron n) {
        double act = n.getActivation();
        if (act >= getUpperBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act + n.getIncrement());
            } else {
                act = act + n.getIncrement();
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    @Override
    public void contextualDecrement(Neuron n) {
        double act = n.getActivation();
        if (act <= getLowerBound() && isClipped()) {
            return;
        } else {
            if (isClipped()) {
                act = clip(act - n.getIncrement());
            } else {
                act = act - n.getIncrement();
            }
            n.setActivation(act);
            n.getNetwork().fireNeuronChanged(n);
        }
    }

    /**
     * @return Returns the decayAmount.
     */
    public double getDecayAmount() {
        return decayAmount;
    }

    /**
     * @param decayAmount The decayAmount to set.
     */
    public void setDecayAmount(final double decayAmount) {
        this.decayAmount = decayAmount;
    }

    /**
     * @return Returns the dedayPercentage.
     */
    public double getDecayFraction() {
        return decayFraction;
    }

    /**
     * @param decayFraction The decayFraction to set.
     */
    public void setDecayFraction(final double decayFraction) {
        this.decayFraction = decayFraction;
    }

    /**
     * @return Returns the relAbs.
     */
    public int getRelAbs() {
        return relAbs;
    }

    /**
     * @param relAbs The relAbs to set.
     */
    public void setRelAbs(final int relAbs) {
        this.relAbs = relAbs;
    }

    @Override
    public boolean getAddNoise() {
        return addNoise;
    }

    @Override
    public void setAddNoise(final boolean addNoise) {
        this.addNoise = addNoise;
    }

    @Override
    public ProbabilityDistribution getNoiseGenerator() {
        return noiseGenerator;
    }

    @Override
    public void setNoiseGenerator(final ProbabilityDistribution noise) {
        this.noiseGenerator = noise;
    }

    /**
     * @return Returns the baseLine.
     */
    public double getBaseLine() {
        return baseLine;
    }

    /**
     * @param baseLine The baseLine to set.
     */
    public void setBaseLine(final double baseLine) {
        this.baseLine = baseLine;
    }

    @Override
    public String getName() {
        return "Decay";
    }

    @Override
    public double getUpperBound() {
        return ceiling;
    }

    @Override
    public double getLowerBound() {
        return floor;
    }

    @Override
    public void setUpperBound(double ceiling) {
        this.ceiling = ceiling;
    }

    @Override
    public void setLowerBound(double floor) {
        this.floor = floor;
    }

    @Override
    public boolean isClipped() {
        return clipping;
    }

    @Override
    public void setClipped(boolean clipping) {
        this.clipping = clipping;
    }

}

File: src/main/kotlin/org/simbrain/network/updaterules/DecayRule.kt
package org.simbrain.network.updaterules

import org.simbrain.network.core.Layer
import org.simbrain.network.core.Network
import org.simbrain.network.core.Neuron
import org.simbrain.network.updaterules.interfaces.ClippedUpdateRule
import org.simbrain.network.updaterules.interfaces.NoisyUpdateRule
import org.simbrain.network.util.EmptyMatrixData
import org.simbrain.network.util.EmptyScalarData
import org.simbrain.util.UserParameter
import org.simbrain.util.propertyeditor.GuiEditable
import org.simbrain.util.stats.ProbabilityDistribution
import org.simbrain.util.stats.distributions.UniformRealDistribution
import kotlin.math.abs

/**
 * **DecayNeuron** implements various forms of standard decay.
 */
open class DecayRule : NeuronUpdateRule<EmptyScalarData, EmptyMatrixData>(), ClippedUpdateRule, NoisyUpdateRule {

    enum class UpdateType {
        Relative, Absolute
    }

    @UserParameter(
        label = "Update Type",
        description = "Relative (percentage decay of current activation) vs. absolute (fixed decay amount)",
        order = 10
    )
    var updateType: UpdateType = UpdateType.Relative

    var decayAmount by GuiEditable(
        initValue = .1,
        label = "Decay amount",
        description = "The amount by which the activation is changed each iteration if absolute decay is used.",
        onUpdate = { enableWidget(widgetValue(DecayRule::updateType) == UpdateType.Absolute) },
        order = 20
    )

    var decayFraction by GuiEditable(
        initValue = .1,
        label = "Decay fraction",
        description = "The proportion of the distance between the current value and the base-line value, "
                + "by which the activation is changed each iteration if relative decay is chosen.",
        onUpdate = { enableWidget(widgetValue(DecayRule::updateType) == UpdateType.Relative) },
        order = 30
    )

    @UserParameter(
        label = "Base Line",
        description = "An option to add noise.",
        increment = .1,
        order = 40)
    var baseLine: Double = 0.0

    /**
     * Clipping.
     */
    override var isClipped: Boolean = true

    /**
     * Noise generator.
     */
    override var noiseGenerator: ProbabilityDistribution = UniformRealDistribution()

    /**
     * Add noise to the neuron.
     */
    override var addNoise: Boolean = false

    /**
     * The upper bound of the activity if clipping is used.
     */
    override var upperBound: Double = DEFAULT_CEILING

    /**
     * The lower bound of the activity if clipping is used.
     */
    override var lowerBound: Double = DEFAULT_FLOOR

    context(Network)
    override fun apply(layer: Layer, dataHolder: EmptyMatrixData) {
        for (i in 0 until layer.activations.nrow()) {
            layer.activations[i, 0] = decayRule(
                layer.inputs[i, 0],
                layer.activations[i, 0]
            )
        }
        clip(layer.activations)
    }

    context(Network)
    override fun apply(neuron: Neuron, data: EmptyScalarData) {
        neuron.activation = decayRule(
            neuron.input,
            neuron.activation
        )
        neuron.clip()
    }

    private fun decayRule(input: Double, activation: Double): Double {
        var retVal = input + activation
        var decayVal = 0.0
        decayVal = if (updateType == UpdateType.Relative) {
            decayFraction * abs(retVal - baseLine)
        } else {
            decayAmount
        }
        if (retVal < baseLine) {
            retVal += decayVal
            if (retVal > baseLine) {
                retVal = baseLine
            }
        } else if (retVal > baseLine) {
            retVal -= decayVal
            if (retVal < baseLine) {
                retVal = baseLine
            }
        }
        if (addNoise) {
            retVal += noiseGenerator.sampleDouble()
        }
        return retVal
    }

    override fun createMatrixData(size: Int): EmptyMatrixData {
        return EmptyMatrixData
    }

    override fun createScalarData(): EmptyScalarData {
        return EmptyScalarData
    }

    override val timeType: Network.TimeType
        get() = Network.TimeType.DISCRETE

    override fun copy(): DecayRule {
        val dn = DecayRule()
        dn.updateType = updateType
        dn.decayAmount = decayAmount
        dn.decayFraction = decayFraction
        dn.isClipped = isClipped
        dn.upperBound = upperBound
        dn.lowerBound = lowerBound
        dn.addNoise = addNoise
        dn.noiseGenerator = noiseGenerator.copy()
        return dn
    }

    override val name: String
        get() = "Decay"

    companion object {
        /**
         * The Default upper bound.
         */
        private const val DEFAULT_CEILING = 1.0

        /**
         * The Default lower bound.
         */
        private const val DEFAULT_FLOOR = -1.0
    }
}